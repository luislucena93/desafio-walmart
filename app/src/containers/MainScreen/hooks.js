import {useEffect, useState} from 'react';

const useFetchProducts = () => {
    const [products, setProducts] = useState([]);

    useEffect(() => {
        fetch('/api/products', {method: 'get'})
            .then((response) => {
                response.json()
                    .then((products) => {
                        setProducts(products);
                    })
            })
    }, []);

    return products;
};

const useGetDiscounts = (brandsTotals) => {
    const [discounts, setDiscounts] = useState({maxDiscountReached: null, maxDiscountPossible: null});
    useEffect(() => {
        fetch('/api/discounts', {
            method: 'post',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify(brandsTotals)
        }).then((response) => {
            response.json()
                .then((discounts) => {
                    setDiscounts(discounts);
                })
        })
    }, [brandsTotals]);
    return discounts;
}

export {useFetchProducts, useGetDiscounts};