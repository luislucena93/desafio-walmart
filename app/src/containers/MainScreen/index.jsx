import React, {Fragment, useState} from 'react';
import Header from '../../components/Header/index.jsx';
import ProductsList from '../../components/ProductsList/index.jsx';
import ShoppingCart from '../../components/ShoppingCart/index.jsx';
import {useFetchProducts, useGetDiscounts} from './hooks';

const MainScreen = () => {
        const [selectedProducts, setSelectedProducts] = useState({});
        const [brandsTotals, setBrandTotals] = useState({});
        const [showShoppingCart, setShowShoppingCart] = useState(false);
        const products = useFetchProducts();
        const discounts = useGetDiscounts(brandsTotals);

        const openShoppingCart = () => {
            setShowShoppingCart(true);
        }

        const closeShoppingCart = () => {
            setShowShoppingCart(false);
        }

        const addProduct = (product) => {
            const productCount = selectedProducts[product.id] !== undefined ? selectedProducts[product.id].count : 0;
            const newProduct = {};
            newProduct[product.id] = {count: productCount + 1, info: product};
            setSelectedProducts(prevState => {
                return ({...prevState, ...newProduct})
            });
            const newBrandsTotals = {};
            newBrandsTotals[product.brand] = (brandsTotals[product.brand] || 0) + product.price;
            setBrandTotals(prevState => {
                return ({...prevState, ...newBrandsTotals});
            });
        };

        const removeProduct = (product) => {
            const productCount = selectedProducts[product.id].count - 1;
            const newSelectedProducts = {...selectedProducts};
            if (productCount === 0) {
                delete newSelectedProducts[product.id];
            } else {
                newSelectedProducts[product.id].count = productCount;
            }
            setSelectedProducts({...newSelectedProducts});
            const productBrandTotal = brandsTotals[product.brand] - product.price;
            const newBrandsTotals = {...brandsTotals};
            if (productBrandTotal === 0) {
                delete newBrandsTotals[product.brand];
            } else {
                newBrandsTotals[product.brand] = productBrandTotal;
            }
            setBrandTotals({...newBrandsTotals});
        };

        return (
            <Fragment>
                <ShoppingCart show={showShoppingCart} closeShoppingCart={closeShoppingCart}
                              selectedProducts={selectedProducts} brandTotals={brandsTotals} discounts={discounts}
                              addProduct={addProduct} removeProduct={removeProduct}/>
                <Header selectedProducts={selectedProducts} openShoppingCart={openShoppingCart}/>
                <ProductsList products={products} addProduct={addProduct} removeProduct={removeProduct}
                              selectedProducts={selectedProducts}/>
            </Fragment>
        )
    }
;

export default MainScreen;