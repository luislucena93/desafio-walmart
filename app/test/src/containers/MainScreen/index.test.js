import React from 'react';
import {render, screen, fireEvent, act} from '@testing-library/react';
import MainScreen from "../../../../../app/src/containers/MainScreen/index";
import * as hooks from '../../../../../app/src/containers/MainScreen/hooks';

describe('MainScreen container', () => {


    const testProducts = [{
        _id: 'testId',
        id: 'testId',
        brand: 'testBrand',
        description: 'testDescription',
        image: 'testImage',
        price: 1000
    }];
    const testDiscounts = {
        maxDiscountReached: {brand: 'testBrand', threshold: 10000, discount: 1000},
        maxDiscountPossible: {brand: 'testBrand', threshold: 10000, discount: 1000}
    };

    beforeAll(() => {
        jest.spyOn(hooks, 'useFetchProducts').mockImplementation(() => {
            return testProducts;
        });
        jest.spyOn(hooks, 'useGetDiscounts').mockImplementation(() => {
            return testDiscounts;
        });
    })

    afterAll(() => {
        jest.clearAllMocks();
    })


    it('should render MainScreen correctly', () => {

        const container = render(<MainScreen/>);
        expect(container).toMatchSnapshot();

    });

    it('should simulate data flow', () => {
        jest.useFakeTimers();
        render(<MainScreen/>);
        act(() => {
            fireEvent.click(screen.getByText("Agregar"));
        });
        expect(screen.getByTestId("totalProducts")).toHaveTextContent("1");
        act(() => {
            fireEvent.click(screen.getByTestId("addProduct"));
        });
        expect(screen.getByTestId("totalProducts")).toHaveTextContent("2");
        act(() => {
            fireEvent.click(screen.getByAltText("shoppingCartIcon"));
        });
        expect(screen.getByTestId("overlay")).toHaveClass("show");
        act(() => {
            fireEvent.click(screen.getByTestId("overlay"));
            jest.runAllTimers();
        });
        expect(screen.getByTestId("overlay")).not.toHaveClass("show");
        act(() => {
            fireEvent.click(screen.getByTestId("removeProduct"));
        });
        expect(screen.getByTestId("totalProducts")).toHaveTextContent("1");
        act(() => {
            fireEvent.click(screen.getByTestId("removeProduct"));
        });
        expect(screen.getByTestId("totalProducts")).toHaveTextContent("0");
        jest.useRealTimers();
    });
});