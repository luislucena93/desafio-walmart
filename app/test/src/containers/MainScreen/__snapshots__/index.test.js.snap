// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen container should render MainScreen correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="overlay "
        data-testid="overlay"
      >
        <div
          class="cart "
          data-testid="cart"
          id="cart"
        >
          <div
            class="cartContent"
          >
            <p
              class="title"
            >
              Resumen
            </p>
            <div
              class="selectedProductsList"
            />
            <div
              class="totals"
            >
              <div
                class="labelAmount"
              >
                <p>
                  Subtotal de productos
                </p>
                <p>
                  $
                  0
                </p>
              </div>
              <div>
                <div
                  class="labelAmount appliedDiscount"
                >
                  <p>
                    Descuento por marca
                  </p>
                  <p>
                    -$
                    1.000
                  </p>
                </div>
                <p
                  class="discountText"
                >
                  Se aplicó un descuento de 
                  <span>
                    $
                    1.000
                  </span>
                   por haber  comprado al menos 
                  <span>
                    $
                    10.000
                  </span>
                   en  productos 
                  testBrand
                </p>
              </div>
              <div
                class="labelAmount total"
              >
                <p>
                  Total a pagar
                </p>
                <p>
                  $
                  -1.000
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="headerBar"
      >
        <img
          alt="Logo"
          class="logo"
          src="test-file-stub"
        />
        <div
          class="shoppingCart"
        >
          <img
            alt="shoppingCartIcon"
            src="test-file-stub"
          />
          <div
            data-testid="totalProducts"
          >
            0
          </div>
        </div>
      </div>
      <div
        class="productsList"
      >
        <div
          class="productDisplay"
        >
          <div>
            <img
              class="thumbnail"
              src="http://testImage"
            />
            <div
              class="separator"
            />
            <p
              class="productDescription"
            >
              <span>
                testBrand
              </span>
               
              testDescription
            </p>
            <p
              class="price"
            >
              $
              1.000
            </p>
          </div>
          <button
            class="addButton"
          >
            Agregar
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="overlay "
      data-testid="overlay"
    >
      <div
        class="cart "
        data-testid="cart"
        id="cart"
      >
        <div
          class="cartContent"
        >
          <p
            class="title"
          >
            Resumen
          </p>
          <div
            class="selectedProductsList"
          />
          <div
            class="totals"
          >
            <div
              class="labelAmount"
            >
              <p>
                Subtotal de productos
              </p>
              <p>
                $
                0
              </p>
            </div>
            <div>
              <div
                class="labelAmount appliedDiscount"
              >
                <p>
                  Descuento por marca
                </p>
                <p>
                  -$
                  1.000
                </p>
              </div>
              <p
                class="discountText"
              >
                Se aplicó un descuento de 
                <span>
                  $
                  1.000
                </span>
                 por haber  comprado al menos 
                <span>
                  $
                  10.000
                </span>
                 en  productos 
                testBrand
              </p>
            </div>
            <div
              class="labelAmount total"
            >
              <p>
                Total a pagar
              </p>
              <p>
                $
                -1.000
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="headerBar"
    >
      <img
        alt="Logo"
        class="logo"
        src="test-file-stub"
      />
      <div
        class="shoppingCart"
      >
        <img
          alt="shoppingCartIcon"
          src="test-file-stub"
        />
        <div
          data-testid="totalProducts"
        >
          0
        </div>
      </div>
    </div>
    <div
      class="productsList"
    >
      <div
        class="productDisplay"
      >
        <div>
          <img
            class="thumbnail"
            src="http://testImage"
          />
          <div
            class="separator"
          />
          <p
            class="productDescription"
          >
            <span>
              testBrand
            </span>
             
            testDescription
          </p>
          <p
            class="price"
          >
            $
            1.000
          </p>
        </div>
        <button
          class="addButton"
        >
          Agregar
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
